platform/linux-generic/odp_packet.c

```diff
@@ -2887,43 +2887,40 @@ uint64_t odp_packet_aging_tmo(odp_packet_t pkt)
 
 int odp_packet_tx_compl_request(odp_packet_t pkt, const odp_packet_tx_compl_opt_t *opt)
 {
-	(void)pkt;
-	(void)opt;
-
-	return -1;
-}
+	odp_packet_hdr_t *pkt_hdr = packet_hdr(pkt);
 
-int odp_packet_has_tx_compl_request(odp_packet_t pkt)
-{
-	(void)pkt;
+	pkt_hdr->p.flags.tx_compl = 1;
+	pkt_hdr->tx_compl_queue = opt->queue;
+	pkt_hdr->tx_compl_mode = opt->mode;
 
 	return 0;
 }
 
-odp_packet_tx_compl_t odp_packet_tx_compl_from_event(odp_event_t ev)
-{
-	(void)ev;
-
-	return ODP_PACKET_TX_COMPL_INVALID;
-}
-
-odp_event_t odp_packet_tx_compl_to_event(odp_packet_tx_compl_t tx_compl)
+int odp_packet_has_tx_compl_request(odp_packet_t pkt)
 {
-	(void)tx_compl;
+	odp_packet_hdr_t *pkt_hdr = packet_hdr(pkt);
 
-	return ODP_EVENT_INVALID;
+	return pkt_hdr->p.flags.tx_compl;
 }
 
 void odp_packet_tx_compl_free(odp_packet_tx_compl_t tx_compl)
 {
-	(void)tx_compl;
+	if (odp_unlikely(tx_compl == ODP_PACKET_TX_COMPL_INVALID)) {
+		ODP_ERR("Bad TX completion event handle\n");
+		return;
+	}
+
+	odp_packet_free(odp_packet_from_event(odp_packet_tx_compl_to_event(tx_compl)));
```

Comment:
At this point event type should be TX_COMPL (could be checked in debug mode) and the type should be reset back to PACKET before the free call, so that the next alloc returns PACKET and not TX_COMPL event.

Reply:
Done in v2.